# 1) Importing data

In [5]:
import numpy as np
import pandas as pd

## 1.a) Import the collection set
The collection set contains metadata of CORD-19 academic papers.

The preprocessed and filtered CORD-19 dataset is available on the Gitlab repository here: https://gitlab.com/checkthat_lab/clef2025-checkthat-lab/-/tree/main/task4/subtask_4b

Participants should first download the file then upload it on the Google Colab session with the following steps.


In [6]:
# 1) Download the collection set from the Gitlab repository: https://gitlab.com/checkthat_lab/clef2025-checkthat-lab/-/tree/main/task4/subtask_4b
# 2) Drag and drop the downloaded file to the "Files" section (left vertical menu on Colab)
# 3) Modify the path to your local file path
PATH_COLLECTION_DATA = 'subtask_4b/subtask4b_collection_data.pkl' #MODIFY PATH

In [7]:
df_collection = pd.read_pickle(PATH_COLLECTION_DATA)

In [8]:
df_collection.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7718 entries, 162 to 1056448
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   cord_uid          7718 non-null   object        
 1   source_x          7718 non-null   object        
 2   title             7718 non-null   object        
 3   doi               7677 non-null   object        
 4   pmcid             4959 non-null   object        
 5   pubmed_id         6233 non-null   object        
 6   license           7718 non-null   object        
 7   abstract          7718 non-null   object        
 8   publish_time      7715 non-null   object        
 9   authors           7674 non-null   object        
 10  journal           6668 non-null   object        
 11  mag_id            0 non-null      float64       
 12  who_covidence_id  528 non-null    object        
 13  arxiv_id          20 non-null     object        
 14  label             7718 n

In [9]:
df_collection.head()

cord_uid source_x                                              title  \
162   umvrwgaw      PMC  Professional and Home-Made Face Masks Reduce E...   
611   spiud6ok      PMC                               The Failure of R (0)   
918   aclzp3iy      PMC  Pulmonary sequelae in a patient recovered from...   
993   ycxyn2a2      PMC  What was the primary mode of smallpox transmis...   
1053  zxe95qy9      PMC  Lessons from the History of Quarantine, from P...   

                               doi       pmcid pubmed_id      license  \
162   10.1371/journal.pone.0002618  PMC2440799  18612429        cc-by   
611            10.1155/2011/527610  PMC3157160  21860658        cc-by   
918        10.4103/0970-2113.99118  PMC3424870  22919170  cc-by-nc-sa   
993       10.3389/fcimb.2012.00150  PMC3509329  23226686        cc-by   
1053        10.3201/eid1902.120312  PMC3559034  23343512        no-cc   

                                               abstract publish_time  \
162   BACKGROUND: Governments are preparing for a po...   2008-07-09   
611   The basic reproductive ratio, R (0), is one of...   2011-08-16   
918   The pandemic of swine flu (H1N1) influenza spr...         2012   
993   The mode of infection transmission has profoun...   2012-11-29   
1053  In the new millennium, the centuries-old strat...   2013-02-03   

                                                authors  \
162   van der Sande, Marianne; Teunis, Peter; Sabel,...   
611       Li, Jing; Blakeley, Daniel; Smith?, Robert J.   
918   Singh, Virendra; Sharma, Bharat Bhushan; Patel...   
993                                   Milton, Donald K.   
1053                                  Tognotti, Eugenia   

                          journal  mag_id who_covidence_id arxiv_id     label  \
162                      PLoS One     NaN              NaN      NaN  umvrwgaw   
611       Comput Math Methods Med     NaN              NaN      NaN  spiud6ok   
918                    Lung India     NaN              NaN      NaN  aclzp3iy   
993   Front Cell Infect Microbiol     NaN              NaN      NaN  ycxyn2a2   
1053             Emerg Infect Dis     NaN              NaN      NaN  zxe95qy9   

           time       timet  
162  2008-07-09  1215561600  
611  2011-08-16  1313452800  
918  2012-01-01  1325376000  
993  2012-11-29  1354147200  
1053 2013-02-03  1359849600

## 1.b) Import the query set

The query set contains tweets with implicit references to academic papers from the collection set.

The preprocessed query set is available on the Gitlab repository here: https://gitlab.com/checkthat_lab/clef2025-checkthat-lab/-/tree/main/task4/subtask_4b

Participants should first download the file then upload it on the Google Colab session with the following steps.

In [10]:
# 1) Download the query tweets from the Gitlab repository: https://gitlab.com/checkthat_lab/clef2025-checkthat-lab/-/tree/main/task4/subtask_4b?ref_type=heads
# 2) Drag and drop the downloaded file to the "Files" section (left vertical menu on Colab)
# 3) Modify the path to your local file path
PATH_QUERY_TRAIN_DATA = 'subtask_4b/subtask4b_query_tweets_train.tsv' #MODIFY PATH
PATH_QUERY_DEV_DATA = 'subtask_4b/subtask4b_query_tweets_dev.tsv' #MODIFY PATH

In [11]:
df_query_train = pd.read_csv(PATH_QUERY_TRAIN_DATA, sep = '\t')
df_query_dev = pd.read_csv(PATH_QUERY_DEV_DATA, sep = '\t')

In [12]:
df_query_train.head()

post_id                                         tweet_text  cord_uid
0        0  Oral care in rehabilitation medicine: oral vul...  htlvpvz5
1        1  this study isn't receiving sufficient attentio...  4kfl29ul
2        2  thanks, xi jinping. a reminder that this study...  jtwb17u8
3        3  Taiwan - a population of 23 million has had ju...  0w9k8iy1
4        4  Obtaining a diagnosis of autism in lower incom...  tiqksd69

In [13]:
df_query_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12853 entries, 0 to 12852
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   post_id     12853 non-null  int64 
 1   tweet_text  12853 non-null  object
 2   cord_uid    12853 non-null  object
dtypes: int64(1), object(2)
memory usage: 301.4+ KB


In [14]:
df_query_dev.head()

post_id                                         tweet_text  cord_uid
0       16  covid recovery: this study from the usa reveal...  3qvh482o
1       69  "Among 139 clients exposed to two symptomatic ...  r58aohnu
2       73  I recall early on reading that researchers who...  sts48u9i
3       93  You know you're credible when NIH website has ...  3sr2exq9
4       96  Resistance to antifungal medications is a grow...  ybwwmyqy

In [15]:
df_query_dev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1400 entries, 0 to 1399
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   post_id     1400 non-null   int64 
 1   tweet_text  1400 non-null   object
 2   cord_uid    1400 non-null   object
dtypes: int64(1), object(2)
memory usage: 32.9+ KB


# 2) BM25 Baseline


In [17]:
#!pip install rank_bm25
from rank_bm25 import BM25Okapi
from tqdm import tqdm

In [18]:
# Create the BM25 corpus
corpus = df_collection[:][['title', 'abstract']].apply(lambda x: f"{x['title']} {x['abstract']}", axis=1).tolist()
cord_uids = df_collection[:]['cord_uid'].tolist()
tokenized_corpus = [doc.split(' ') for doc in corpus]
bm25 = BM25Okapi(tokenized_corpus)

In [19]:
text2bm25top = {}
def get_top_cord_uids(query):
  if query in text2bm25top.keys():
      return text2bm25top[query]
  else:
      tokenized_query = query.split(' ')
      doc_scores = bm25.get_scores(tokenized_query)
      indices = np.argsort(-doc_scores)[:100]
      bm25_topk = [cord_uids[x] for x in indices]

      text2bm25top[query] = bm25_topk
      return bm25_topk


In [16]:
# Retrieve topk candidates using the BM25 model
# df_query_train['bm25_topk'] = df_query_train['tweet_text'].apply(lambda x: get_top_cord_uids(x))
# df_query_dev['bm25_topk'] = df_query_dev['tweet_text'].apply(lambda x: get_top_cord_uids(x))

In [17]:
# Retrieve topk candidates using the BM25 model
bm25_results = []
for text in tqdm(df_query_train['tweet_text']):
    bm25_results.append(get_top_cord_uids(text))

df_query_train['bm25_topk'] = bm25_results

bm25_results = []
for text in tqdm(df_query_dev['tweet_text']):
    bm25_results.append(get_top_cord_uids(text))

df_query_dev['bm25_topk'] = bm25_results

  0%|          | 28/12853 [00:05<39:33,  5.40it/s]


KeyboardInterrupt: 

# 2.2) Execution of BM25 on modified tweets

In [21]:
# In case of compatibility issues between spacy and numpy - run the following reinstallation and restart the runtime of the colab
# !pip uninstall -y numpy spacy
# !pip install numpy==1.23.5
# !pip install spacy==3.4.4
# !python -m spacy download en_core_web_sm

Found existing installation: numpy 2.2.4
Uninstalling numpy-2.2.4:
  Successfully uninstalled numpy-2.2.4
Found existing installation: spacy 3.4.4
Uninstalling spacy-3.4.4:
  Successfully uninstalled spacy-3.4.4
  Using cached numpy-1.23.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.3 kB)
Using cached numpy-1.23.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.1 MB)
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
scispacy 0.5.5 requires spacy<3.8.0,>=3.7.0, which is not installed.
fastai 2.7.19 requires spacy<4, which is not installed.
xarray 2025.1.2 requires numpy>=1.24, but you have numpy 1.23.5 which is incompatible.
bigframes 1.42.0 requires numpy>=1.24.0, but you have numpy 1.23.5 which is incompatible.
chex 0.1.89 requires numpy>=1.24.1, but you have numpy 1.23.5 which is incompatible.
albumentations 2.0.5 requi

  Using cached spacy-3.4.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (24 kB)
Using cached spacy-3.4.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.4 MB)
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
en-ner-bionlp13cg-md 0.5.4 requires spacy<3.8.0,>=3.7.4, but you have spacy 3.4.4 which is incompatible.
scispacy 0.5.5 requires spacy<3.8.0,>=3.7.0, but you have spacy 3.4.4 which is incompatible.
2025-04-16 07:45:00.300225: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1744789500.328834   35297 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1744789500.337593   35297 cuda_bla

Import required modules (use en_ner_bionlp13cg_md model that focuses on biomedical named entity recognition).
Download nltk Wordnet

In [17]:
# !pip install -q spacy scispacy https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.4/en_ner_bionlp13cg_md-0.5.4.tar.gz

  Preparing metadata (setup.py) ... done


In [23]:
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

Now import the required libraries

In [24]:
import re
import spacy
from nltk.corpus import wordnet as wn

nlp = spacy.load("en_ner_bionlp13cg_md")

/usr/local/lib/python3.11/dist-packages/spacy/language.py:2195: FutureWarning: Possible set union at position 6328
  deserializers["tokenizer"] = lambda p: self.tokenizer.from_disk(  # type: ignore[union-attr]


Define tweets processing functions

In [21]:
# lowercase, remove mentions and hashtags
def clean_tweet(text):
    text = text.lower()
    text = re.sub(r"@\w+", "", text)
    text = re.sub(r"#\w+", "", text)
    return text.strip()

# extract scientific terms using SciSpacy
def extract_terms(text):
    doc = nlp(text)
    return list(set(ent.text for ent in doc.ents))

# extract WordNet synonyms (not specific to biomedical)
def get_synonyms(term):
    synonyms = set()
    for syn in wn.synsets(term.replace(" ", "_")):
        for lemma in syn.lemmas():
            synonyms.add(lemma.name().replace("_", " "))
    return list(synonyms)

# reformulate a tweet: clean, extract terms, expand with synonyms
def reformulate(tweet):
    cleaned = clean_tweet(tweet)
    terms = extract_terms(cleaned)
    synonyms = []
    for term in terms:
        synonyms.extend(get_synonyms(term))
    expansion = ' '.join(set(synonyms))
    return f"{cleaned} {expansion}".strip()


Prepare the dataframes

In [25]:
tqdm.pandas()
df_query_train['expanded_tweet_text'] = df_query_train['tweet_text'].progress_apply(reformulate)
df_query_dev['expanded_tweet_text'] = df_query_dev['tweet_text'].progress_apply(reformulate)

100%|██████████| 1400/1400 [00:21<00:00, 66.43it/s]


Observe the expanded tweets

In [26]:
def wrapped_print(tweet):
    for j in range(0, len(tweet), 120):
        print(tweet[j:j+100])

for i, (tweet, expanded_tweet) in enumerate(zip(df_query_train['tweet_text'].head(5), df_query_train['expanded_tweet_text'].head(5)), 1):
    print(f"\n--- Tweet {i} ---\n")
    wrapped_print(tweet)
    print(f"\n--- Expanded Tweet {i} ---\n")
    wrapped_print(expanded_tweet)


--- Tweet 1 ---

Oral care in rehabilitation medicine: oral vulnerability, oral muscle wasting, and hospital-associat

--- Expanded Tweet 1 ---

oral care in rehabilitation medicine: oral vulnerability, oral muscle wasting, and hospital-associat
oral exam oral examination unwritten viva voce oral

--- Tweet 2 ---

this study isn't receiving sufficient attention. it reveals black/latino/indigenous individuals aren
o covid at higher rates than whites but are also passing away *earlier*.   90% of white fatalities o
0% of black fatalities occur in those 55+,  89% of indigenous fatalities occur in those 45+

--- Expanded Tweet 2 ---

this study isn't receiving sufficient attention. it reveals black/latino/indigenous individuals aren
o covid at higher rates than whites but are also passing away *earlier*.   90% of white fatalities o
0% of black fatalities occur in those 55+,  89% of indigenous fatalities occur in those 45+

--- Tweet 3 ---

thanks, xi jinping. a reminder that this study con

In [27]:
bm25_train_results = []
bm25_dev_results = []

In [28]:
# Retrieve topk candidates using the BM25 model
for text in tqdm(df_query_train['expanded_tweet_text']):
    bm25_train_results.append(get_top_cord_uids(text))

df_query_train['bm25_expanded_topk'] = bm25_train_results

for text in tqdm(df_query_dev['expanded_tweet_text']):
    bm25_dev_results.append(get_top_cord_uids(text))

df_query_dev['bm25_expanded_topk'] = bm25_dev_results

100%|██████████| 1400/1400 [03:39<00:00,  6.37it/s]


In [31]:
df_query_train['bm25_expanded_topk'].shape
df_query_dev['bm25_expanded_topk'].shape

(1400,)

# 3) Evaluating BM25 baseline with MRR@5

In [32]:
# Evaluate retrieved candidates using MRR@k
def get_performance_mrr(data, col_gold, col_pred, list_k = [1, 5, 10]):
    d_performance = {}
    for k in list_k:
        data["in_topx"] = data.apply(lambda x: (1/([i for i in x[col_pred][:k]].index(x[col_gold]) + 1) if x[col_gold] in [i for i in x[col_pred][:k]] else 0), axis=1)
        #performances.append(data["in_topx"].mean())
        d_performance[k] = data["in_topx"].mean()
    return d_performance

In [38]:
results_train = get_performance_mrr(df_query_train, 'cord_uid', 'bm25_topk')
results_dev = get_performance_mrr(df_query_dev, 'cord_uid', 'bm25_topk')
# Printed MRR@k results in the following format: {k: MRR@k}
print(f"Results on the train set: {results_train}")
print(f"Results on the dev set: {results_dev}")

Results on the train set: {1: np.float64(0.5081303975725512), 5: np.float64(0.5509777224513084), 10: np.float64(0.5559614579512657)}
Results on the dev set: {1: np.float64(0.5057142857142857), 5: np.float64(0.5522738095238094), 10: np.float64(0.557658163265306)}


# 3.2) Evaluating expanded BM25 with MRR@5

In [33]:
results_train = get_performance_mrr(df_query_train, 'cord_uid', 'bm25_expanded_topk')
results_dev = get_performance_mrr(df_query_dev, 'cord_uid', 'bm25_expanded_topk')
# Printed MRR@k results in the following format: {k: MRR@k}
print(f"Expanded BM25: Results on the train set: {results_train}")
print(f"Expanded BM25: Results on the dev set: {results_dev}")

Expanded BM25: Results on the train set: {1: 0.4654166342488135, 5: 0.5104865271402267, 10: 0.5160982118929679}
Expanded BM25: Results on the dev set: {1: 0.4657142857142857, 5: 0.5149404761904761, 10: 0.5205538548752834}


# 4) Exporting results to prepare the submission on Codalab

In [ ]:
df_query_dev['preds'] = df_query_dev['bm25_topk'].apply(lambda x: x[:5])

In [ ]:
df_query_dev[['post_id', 'preds']].to_csv('predictions.tsv', index=None, sep='\t')

# Results documentation:
#### Baseline BM25:
Results on the train set: {1: 0.5081303975725512, 5: 0.5509777224513084, 10: np.0.5559614579512657}\
Results on the dev set: {1: 0.5057142857142857, 5: 0.5522738095238094, 10: 0.557658163265306}
#### Expanded BM25: (cleaned and enriched tweets)
Results on the train set: {1: 0.4654166342488135, 5: 0.5104865271402267, 10: 0.5160982118929679}\
Results on the dev set: {1: 0.4657142857142857, 5: 0.5149404761904761, 10: 0.5205538548752834}
